protocol/x/affiliates/keeper/keeper_test.go

```diff
@@ -14,3 +20,193 @@ func TestLogger(t *testing.T) {
 	logger := tApp.App.AffiliatesKeeper.Logger(ctx)
 	require.NotNil(t, logger)
 }
+
+func TestRegisterAffiliate(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+	k := tApp.App.AffiliatesKeeper
+
+	tests := []struct {
+		name        string
+		referee     string
+		affiliate   string
+		expectError bool
+	}{
+		{
+			name:        "Register new affiliate",
+			referee:     "referee1",
+			affiliate:   "affiliate1",
+			expectError: false,
+		},
+		{
+			name:        "Register existing referee",
+			referee:     "referee1",
+			affiliate:   "affiliate2",
+			expectError: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			err := k.RegisterAffiliate(ctx, tc.referee, tc.affiliate)
+			if tc.expectError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				affiliate, exists := k.GetReferredBy(ctx, tc.referee)
+				require.True(t, exists)
+				require.Equal(t, tc.affiliate, affiliate)
+			}
+		})
+	}
+}
+
+func TestAddReferredVolume(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+	k := tApp.App.AffiliatesKeeper
+
+	affiliate := "affiliate1"
+	initialVolume := dtypes.NewInt(1000)
+	addedVolume := dtypes.NewInt(500)
+
+	err := k.AddReferredVolume(ctx, affiliate, initialVolume)
+	require.NoError(t, err)
+
+	volume, exists := k.GetReferredVolume(ctx, affiliate)
+	require.True(t, exists)
+	require.Equal(t, initialVolume, volume)
+
+	err = k.AddReferredVolume(ctx, affiliate, addedVolume)
+	require.NoError(t, err)
+
+	updatedVolume, exists := k.GetReferredVolume(ctx, affiliate)
+	require.True(t, exists)
+	require.Equal(t, dtypes.NewInt(1500), updatedVolume)
+}
+
+func TestGetTakerFeeShareViaReferredVolume(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+	k := tApp.App.AffiliatesKeeper
+
+	// Set up affiliate tiers
+	affiliateTiers := types.AffiliateTiers{
+		Tiers: []types.AffiliateTiers_Tier{
+			{Level: 0, ReqReferredVolume: 0, TakerFeeSharePpm: 1000, ReqStakedWholeCoins: 200},
```

Comment:
Add test case with multiple tiers (for simplicity, just declare `DefaultAffiliateTiers` constant from this [table](https://www.notion.so/dydx/Growth-MVP-MVP-affiliates-program-9001cea80d524d5baa6cbafb830bc278?pvs=4#f447108d68fc47ec899e578dd9c65c25), and use in tests.